crates/red_knot/src/lint.rs

```diff
@@ -140,6 +142,66 @@ fn lint_unresolved_imports(context: &SemanticLintContext) -> QueryResult<()> {
     Ok(())
 }
 
+fn lint_bad_overrides(context: &SemanticLintContext) -> QueryResult<()> {
+    // TODO possibly we should have a special marker on the real typing module (from typeshed) so
+    // if you have your own "typing" module in your project, we don't consider it THE typing module
+    let Some(typing_module) = context.db.resolve_module(ModuleName::new("typing"))? else {
+        debug_assert!(false, "typing module should always exist");
+        return Ok(());
+    };
+    let typing_file = context.db.module_to_file(typing_module)?;
+    let typing_table = context.db.symbol_table(typing_file)?;
+    let Some(typing_override) = typing_table.root_symbol_id_by_name("override") else {
+        debug_assert!(false, "typing.override should exist");
+        return Ok(());
+    };
+
+    // TODO we should maybe index definitions by type instead of iterating all, or else iterate all
+    // just once, match, and branch to all lint rules that care about a type of definition
+    for (symbol, definition) in context.symbols().all_definitions() {
+        if !matches!(definition, Definition::FunctionDef(_)) {
+            continue;
+        }
+        let ty = context
+            .db
+            .infer_definition_type(context.file_id, symbol, definition.clone())?;
+        let Type::Function(func) = ty else {
+            debug_assert!(false, "type of a FunctionDef should always be a Function");
+            continue;
+        };
+        let Some(class) = func.get_containing_class(context.db)? else {
+            // not a method of a class
+            continue;
+        };
+        let mut override_decorated = false;
+        for deco_ty in func.function(context.db)?.decorators() {
+            let Type::Function(deco_func) = deco_ty else {
+                continue;
+            };
+            if deco_func.file() == typing_file && deco_func.symbol(context.db)? == typing_override {
```

Comment:
Yes, I agree this is a good direction to push the API.

I think we can provide APIs that prevent a lint rule from _needing_ to get another file's symbol table. I'm not sure how we prevent a lint rule from being _able_ to do that, though. It seems like it will require not giving lint rules direct db access at all. Which may be feasible but will require a lot of duplication of parts of db API surface on lint-context.

Reply:
I added a GlobalSymbolId type and a `resolve_global_symbol` API that returns it, as well as changing other queries (`infer_symbol_type`, `infer_definition_type`) to take it.

We still use local SymbolIds some places, but fewer now.

This doesn't really do anything to make it harder for a lint rule to get access to some other module's symbol table, but it provides an alternative API that should mean they don't need to.